src/software/ai/evaluation/keep_away.cpp

```diff
@@ -0,0 +1,59 @@
+#include "software/ai/evaluation/keep_away.h"
+
+#include "software/ai/passing/cost_function.h"
+#include "software/math/math_functions.h"
+#include "software/optimization/gradient_descent_optimizer.h"
+
+double ratePasserPointForKeepAway(const Pass& pass, const Team& enemy_team)
+{
+    // the default values for these passing parameters
+    // TODO: cleanup passing parameters as part of #1987
+    // TODO: tweak all of these constants to get better behaviour
+    static constexpr double PASSER_ENEMY_PROXIMITY_IMPORTANCE   = 1.5;
+    static constexpr double RECEIVER_ENEMY_PROXIMITY_IMPORTANCE = 0.;
+    static const auto ENEMY_REACTION_TIME = Duration::fromSeconds(0);
+
+
+    return ratePassEnemyRisk(enemy_team, pass, ENEMY_REACTION_TIME,
+                             RECEIVER_ENEMY_PROXIMITY_IMPORTANCE) *
+           (1 - calculateProximityRisk(pass.passerPoint(), enemy_team,
+                                       PASSER_ENEMY_PROXIMITY_IMPORTANCE));
+}
+
+
+Point findKeepAwayTargetPoint(const World& world, const Pass& best_pass_so_far)
+{
+    // TODO: tweak this constant
+    static constexpr auto KEEPAWAY_SEARCH_CIRCLE_RADIUS = 0.75;
+
+    // the width of both the field boundary sigmoid and the circular search region sigmoid
+    static constexpr auto SIGMOID_WIDTH = 0.05;
+
+    static constexpr auto GRADIENT_STEPS_PER_ITER = 2;
```

Comment:
can you measure how long these two iterations take and whether it causes a significant (>1ms) change in tick time?

Reply:
the entire `findKeepAwayTargetPoint` function takes around 100 microseconds, so this is negligible